src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+        __ subq(rScratch, Address(rsp, SAVED_HAYSTACK_STACK_OFFSET));
+        __ addq(rScratch, rTmp);
+        __ movq(r11, rScratch);
+        __ jmp(L_checkRangeAndReturn);
+
+        __ bind(L_compareFull);
+
+        // rScratch has n - k.  Compare entire string word-by-word
+        __ xorq(r11, r11);
+        __ movq(r10, rScratch);
+        __ jmpb(doCompare);
+
+        __ bind(topLoop);
+        __ addq(r11, 2);
+        __ cmpq(r11, rScratch);
+        __ jg(L_returnRBP);
+
+        __ bind(doCompare);
+        __ leaq(r9, Address(haystack, r11));
+        __ leaq(r12, Address(needle, 0));
+        __ movq(r13, origNeedleLen);
+
+        __C2 arrays_equals(false, r9, r12, r13, rax, rdx, xmm_tmp3, xmm_tmp4, false /* char */, knoreg,
+                           true /* expand_ary2 */);
+        __ testq(rax, rax);
+        __ jz(topLoop);
+
+        // Match found
+        __ jmp(L_checkRangeAndReturn);
+      }
+    }
+
+  } else {  // SSE version
+    assert(false, "Only supports AVX2");
+  }
+
+  return;
+}
+
+// Helper for broadcasting needle elements to ymm registers for compares
+//
+// For UTF-16 encoded needles, broadcast a word at the proper offset to the ymm
+// register (case UU)
+// For the UTF-16 encoded haystack with Latin1 encoded needle (case UL) we have
+// to read into a temp register to zero-extend the single byte needle value, then
+// broadcast words to the ymm register.
+//
+// Parameters:
+// sizeKnown - True if size known at compile time
+// size - the size of the needle.  Pass 0 if unknown at compile time
+// bytesToCompare - see NUMBER_OF_NEEDLE_BYTES_TO_COMPARE above
+// needle - the address of the first byte of the needle
+// needleLen - length of needle if !sizeKnown
+// isUU and isUL - true if argument encoding is UU or UL, respectively
+// _masm - Current MacroAssembler instance pointer
+static void broadcast_additional_needles(bool sizeKnown, int size, int bytesToCompare, Register needle,
+                                         Register needleLen, Register rTmp, bool isUU, bool isUL,
+                                         MacroAssembler *_masm) {
+  Label L_done;
+
+  const XMMRegister byte_1 = XMM_BYTE_1;
+  const XMMRegister byte_2 = XMM_BYTE_2;
+
+  assert_different_registers(needle, needleLen, rTmp);
+
+  bool isU = (isUU || isUL);
+
+  size = sizeKnown ? size : NUMBER_OF_CASES + 1;
+
+  if (bytesToCompare > 2) {
+    if (!sizeKnown) {
+      __ cmpq(needleLen, (isU ? 4 : 2));
+      __ jl_b(L_done);
+    }
+
+    if (size > (isU ? 4 : 2)) {
+      // Add compare for second byte
+      if (isUU) {
+        __ vpbroadcastw(byte_1, Address(needle, 2), Assembler::AVX_256bit);
+      } else if (isUL) {
+        __ movzbl(rTmp, Address(needle, 1));
+        __ movdl(byte_1, rTmp);
+        // 1st byte of needle in words
+        __ vpbroadcastw(byte_1, byte_1, Assembler::AVX_256bit);
+      } else {
+        __ vpbroadcastb(byte_1, Address(needle, 1), Assembler::AVX_256bit);
+      }
+    }
```

Comment:
It will be good to have a function which broadcasts a needle element from a given offset into a vector register.
That function could take (needle address, offset, outout vector register, temps).
Such a function could then be called twice from here and from main function for offset 0.